### ISSUE TITLE
Run as k8s jobs rather than as a single running pod. 

### ISSUE THREAD

**Issue by JacobWeyer** [upbound/provider-terraform#189]
### What problem are you facing?

I'd like to see this operator function in a way where it would spin up workspace runs in parallel for every request (up to a max parallel limit that can be set by the user).

The intention is to use this for developer environments, load testing, integration testing and more in a very dynamic manner. The way the current operator seems to work is sequential by nature and ends up being slower as a result. 

### How could Official Terraform Provider help solve your problem?
By being able to run our terraform we can take advantage of crossplane's flexibility combined with helm and spin up a significant number of micro services and environments very quickly if they can run in parallel rather than being forced to wait for sequential execution. Sequential execution is a real bummer when we have something like RDS or DMS that can take up to 15 minutes to start up properly.

**Comment by bobh66** [comment]
@JacobWeyer the default for the provider is to run one reconciliation at a time, but this is configurable using `--max-reconcile-rate` argument in a `ControllerConfig` and you can set it as high as you want.  See https://github.com/upbound/provider-terraform/blob/main/examples/install.yaml for an example `ControllerConfig`.

However, since the underlying code runs the terraform CLI ,the pod will attempt to use as many CPUs as it has threads configured, so to get a "true" parallel execution you would need to make sure that there are the same number of CPUs as you set for the reconcile rate.

**Comment by JacobWeyer** [comment]
I guess I'm confused why this was designed to run as a single instance instead of having the operator trigger each run as its own job in a similar manner to how something like Github Actions works.

**Comment by bobh66** [comment]
Crossplane providers are designed to be reconciling kubernetes controllers which are responsible for maintaining the state specified in the `spec` field of the resource manifest.  That is a different paradigm than a job dispatcher.

If each CLI command was dispatched as individual jobs they could take advantage of idle CPU resources on other workers but each would still require 1 CPU to run to completion, and it would add complexity to track the remote job completion so that subsequent reconciliations don't run while there is already a process running.

**Comment by balu-ce** [comment]
@bobh66 / @JacobWeyer can we do it in a master slave , where master has configuration . and making slave as replicas which is scalable. will it work ?.

**Comment by JacobWeyer** [comment]
Yeah that makes sense @bobh66, I'm still curious if there's a more distributed batching methodology that'd be beneficial. Especially at scale without just running more jobs in parallel on a single operator.

**Comment by negz** [comment]
I'm a little wary of this idea. Mostly in that I'm wary of `provider-terraform` diverging from how all the other Crossplane providers work.

**Comment by jeanduplessis** [comment]
⚠️ Please note that this repo has migrated to https://github.com/crossplane-contrib/provider-terraform 

Due to GitHub's limitations on transferring issues across organizations, this issue will be closed. 
Please re-open your issue upstream if you still want it considered.

Note that Upbound will no longer actively invest in this provider, but it will endeavor to continue supporting community contributions and publishing new releases as needed.

For users interested in using newer Terraform versions, please note that provider-terraform will NOT support versions of Terraform under the BUSL license (anything after 1.5.x). We encourage users who need newer capabilities to consider https://github.com/upbound/provider-opentofu.
